`-q javaToolchains` is not compatible with configuration cache

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

<!--- Provide a brief summary of the issue in the title above -->

### Expected Behavior
<!--- Tell us what should happen -->
It should be compatible or at least not fail ungracefully

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
It fails

### Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Run `gradle -q javaToolchains` twice in a gradle build with configuration caching enabled, you will see this failure the second try

```
gradle -q javaToolchains

FAILURE: Build failed with an exception.

* What went wrong:
Could not load the value of field `specs` of `org.gradle.api.specs.AndSpec` bean found in task `:javaToolchains` of type `org.gradle.jvm.toolchain.internal.task.ShowToolchainsTask`.
> org.gradle.jvm.toolchain.internal.task.ShowToolchainsTask$$Lambda$1901/0x00000008024fd410

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

CONFIGURE FAILED in 876ms
```

### Steps to Reproduce 
<!---
Provide a self-contained example project as an attached archive or a Github project.
You can use the following template with a Gradle GitHub action set up to showcase your problem: https://github.com/gradle/gradle-issue-reproducer
In the rare cases where this is infeasible, we will also accept a detailed set of instructions.
-->
See above

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL: N/A

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheJavaToolchainsIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheJavaToolchainsIntegrationTest.groovy
@@ -1,0 +1,53 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.gradle.integtests.fixtures.configurationcache.ConfigurationCacheBuildOperationsFixture
+
+class ConfigurationCacheJavaToolchainsIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    protected ConfigurationCacheBuildOperationsFixture configurationCache
+
+    def setup() {
+        configurationCache = newConfigurationCacheFixture()
+    }
+
+    def "show toolchains task works properly"() {
+        when:
+        configurationCacheRun "javaToolchains"
+
+        then:
+        assertStateStored()
+        result.assertTasksExecuted(":javaToolchains")
+
+        when:
+        configurationCacheRun "javaToolchains"
+
+        then:
+        assertStateLoaded()
+        result.assertTasksExecuted(":javaToolchains")
+    }
+
+    private void assertStateStored() {
+        configurationCache.assertStateStored()
+    }
+
+    private void assertStateLoaded() {
+        configurationCache.assertStateLoaded()
+    }
+
+}

--- a/subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/task/ShowToolchainsTask.java
+++ b/subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/task/ShowToolchainsTask.java
@@ -36,6 +36,7 @@ import java.util.List;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
+import static org.gradle.api.internal.lambdas.SerializableLambdas.spec;
 import static org.gradle.internal.logging.text.StyledTextOutput.Style.Description;
 import static org.gradle.internal.logging.text.StyledTextOutput.Style.Identifier;
 import static org.gradle.internal.logging.text.StyledTextOutput.Style.Normal;
@@ -50,7 +51,7 @@ public class ShowToolchainsTask extends DefaultTask {
     private final ToolchainReportRenderer toolchainRenderer = new ToolchainReportRenderer();
 
     public ShowToolchainsTask() {
-        getOutputs().upToDateWhen(element -> false);
+        getOutputs().upToDateWhen(spec(element -> false));
     }
 
     @TaskAction

